spanlogger/spanlogger.go

```diff
@@ -68,22 +75,57 @@ func FromContext(ctx context.Context, fallback log.Logger, resolver TenantResolv
 	if !ok {
 		logger = fallback
 	}
+	sampled := false
 	sp := opentracing.SpanFromContext(ctx)
 	if sp == nil {
 		sp = opentracing.NoopTracer{}.StartSpan("noop")
+	} else {
+		_, sampled = tracing.ExtractSampledTraceID(ctx)
 	}
-	lwc, sampled := withContext(ctx, logger, resolver)
 	return &SpanLogger{
-		Logger:  lwc,
-		Span:    sp,
-		sampled: sampled,
+		ctx:          ctx,
+		baseLogger:   logger,
+		resolver:     resolver,
+		Span:         sp,
+		sampled:      sampled,
+		debugEnabled: debugEnabled(logger),
 	}
 }
 
+// Detect whether we should output debug logging.
+// false iff the logger says it's not enabled; true if the logger doesn't say.
+func debugEnabled(logger log.Logger) bool {
+	if x, ok := logger.(interface{ DebugEnabled() bool }); ok && !x.DebugEnabled() {
+		return false
+	}
+	return true
+}
+
 // Log implements gokit's Logger interface; sends logs to underlying logger and
 // also puts the on the spans.
 func (s *SpanLogger) Log(kvps ...interface{}) error {
-	s.Logger.Log(kvps...)
+	if s.logger == nil {
+		s.logger = s.makeLogger()
+	}
+	s.logger.Log(kvps...)
+	return s.spanLog(kvps...)
+}
+
+// DebugLog is more efficient than level.Debug().Log().
+func (s *SpanLogger) DebugLog(kvps ...interface{}) error {
+	if s.debugEnabled {
+		if s.logger == nil {
+			s.logger = s.makeLogger()
+		}
+		// The call to Log() through an interface makes its argument escape, so make a copy here,
+		// in the debug-only path, so the function is faster for the non-debug path.
+		localCopy := append([]any{}, kvps...)
+		level.Debug(s.logger).Log(localCopy...)
+	}
+	return s.spanLog(kvps...)
```

Comment:
Do we need to add a `level=debug` pair here in the call to `spanLog`? Otherwise events that previously included the level when using `level.Debug().Log()` will not have it when `DebugLog()` is used.

Reply:
I can't say I've ever needed that. Is it something you rely on?

Reply:
Not something I rely on, but it is a change. I'm happy to go ahead without it and see if anyone complains / anything breaks.